third_party/cpu/lib/TritonCPUTransforms/ConvertDotProduct.cpp

```diff
@@ -240,17 +247,187 @@ struct ConvertMulSumToDotHorizontalSum
   }
 };
 
+struct ConvertMulSumToDotPack
+    : public OpRewritePattern<vector::MultiDimReductionOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::MultiDimReductionOp op,
+                                PatternRewriter &rewriter) const override {
+    MLIRContext *ctx = op.getContext();
+    Location loc = op.getLoc();
+    Value acc = op.getAcc();
+    auto resTy = cast<VectorType>(op.getType());
+
+    Value matInput;
+    Value vecInput;
+
+    bool isMatch = isBf16DotProduct(op, false, matInput, vecInput, rewriter);
+    if (!isMatch)
+      return failure();
+
+    // Once we get the matrix input (NxK) and vector input (K),
+    // where N is the output channel dimension
+    // and K is the reduction dimension.
+    // We will generate the following code to perform the dot product.
+    // We will first transpose the matrix so that the output channel dimension
+    // is continuous, so we can store multiple output channels in one
+    // SIMD register.
+    // Then we will loop over the K dimension.
+    // For each iteration over K, we will pull 2 bf16 from the input vector.
+    // Inside the K loop, we will also iterate over the output channels.
+    // For each iteration over the output channel, we will pull
+    // 4 output channel (each containing 2 bf16).
+    // Then we will broadcast the 2 bf16 from the input vector,
+    // dot product it with the 4 output channels (each containing 2 bf16),
+    // and accumulate it with 4 outputs.
+    // We will iterate over N until all output channels are processed.
+    // Then we will move to the next 2 bf16 from the input vector (the K loop).
+    // We will also share the vector input across the output channels.
+    // For example, if we dot product a size 8x8 matrix with a size 8 vector,
+    // the generated pseudo code will be:
+    // matrix = shapecast(matrix, 8x4x2)
+    // matrix = tranpose(matrix, 1, 0, 2) : 8x4x2xbf16 -> 4x8x2xbf16
```

Comment:
Nit dim names in the comment for all these lines to make it clearer
```suggestion
    // matrix = tranpose(matrix, 1, 0, 2) : 8x4x2xbf16 -> 4x8x2xbf16 // KxNxk -> NxKxk?
```

Reply:
Added